ionic-team/stencil
# fix(style): fixes to watching nested and multiple styles on Stencil components

This PR contains a set of small optimizations and bugfixes. I originally wanted to keep them in separate PRs but then found that they were small enough to be combined in one. The following issues get addressed:

## Multiple style urls aren't recognised in watch mode

fixes https://github.com/ionic-team/stencil/issues/5251

When a component is bundled with multiple css files attached to it, the output will look as following:

```ts
const Button = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
    }
    render() {
        return h("button", { class: "button" }, "Button Text ", h("span", null, "OK!"));
    }
};
Button.style = buttonBackgroundCss + buttonTextCss;
```

From this Stencil creates a style tag within the shadow DOM to enable style HMR and injects the combined styles of the component `cmp.style`. Now when we update one CSS file, we push the update to the build context `buildCtx.stylesUpdated` which will cause in replacing the content with what is in `stylesUpdated`. Unfortunately, when we change one file, we only get the update content of that file and not the combined styles.

The patch stores the transform results of styles and is able to send a concatenation of them to the application.

Changes were applied in `src/compiler/bundle/ext-transforms-plugin.ts`

#### Review Guide

To review this change:

- run `git clone https://github.com/a7medm7med/stencil-bug`
- `cd stencil-bug && npm i`
- `npm start`
- in `src/components/button/button.text.scss` chnage color from `red` to e.g. `blue` and verify that: background color is reseted and only styles from `src/components/button/button.text.scss` are applied to the page

to verify fix:

- install dev build: `npm i @stencil/core@4.9.1-dev.1705446396.3c9fd4d`
- restart dev server
- change color in `src/components/button/button.text.scss` and verify that background color remains red 🎉 

## nested files are not applied in watch mode

fixes #2635
fixes #2741
fixes #1908
fixes #1388

it also:
closes #1909 (attempt to fix this bug in a different way which turned out to be incorrect)

If a component references an external style, e.g.:

```ts
@Component({
  tag: 'my-component',
  styleUrl: 'my-component.css',
  shadow: true,
})
export class MyComponent {
  // ...
}
```

or a global style is defined which imports another sass file:

```css
// my-component.css
@import "./foo.scss";

div {
  background-color: var(--body-bg);
}
```
```scss
:host {
  --body-bg: blue;
}
```

then changes in `foo.scss` will not be recognised and user has to restart the dev server.

Changes applied in `src/compiler/plugin/plugin.ts` fix this problem. They were adopted from #3110 and basically just add detected file dependencies to the transform result.

#### Review Guide

To review this change:

- run `git clone git@github.com:emily-ut/stencil-playground.git`
- `cd stencil-playground`
- `git checkout bug-scss-watcher`
- `npm i`
- `npm start`
- change css variable in `src/global/theme/theme.wireframe.scss` to a different color, verify that background color doesn't change

to verify fix:

- install dev build: `npm i @stencil/core@4.9.1-dev.1705446396.3c9fd4d`
- restart dev server
- change css variable in `src/global/theme/theme.wireframe.scss` to a different color, verify that background now gets updated correctlt 🎉 

## Styles outside of project dir aren't picked up by watcher

Currently we add only the files that are within `config.srcDir` to our watch process. If a file is outside of that directory, it won't get picked up. As a fix I added a check to the `extTransformsPlugin` which touches all styles that are used by any of the components and will add a style if the path is relative or absolute and not within the `srcDir`.

Some considerations:
- there might be a better place to add a style to the watch list
- we might not want to tweak this at runtime and rather have the user specify other watch folders

#### Review Guide

To review this change:

- run `git clone git@github.com:James-Wilkinson-git/stencil-bug.git`
- `cd ./stencil-bug/stencil-install/stencil`
- `npm i`
- `npm start`
- change color in `./stencil-bug/shared-style/my-component.scss` to a different color, verify that background color doesn't change

to verify fix:

- `cd ./stencil-bug/stencil-install/stencil`
- install dev build: `npm i @stencil/core@4.9.1-dev.1705446396.3c9fd4d`
- restart dev server
- change color in `./stencil-bug/shared-style/my-component.scss` to a different color, verify that background now gets updated correctlt 🎉
